packages/tpetra/core/src/Tpetra_FEMultiVector_decl.hpp

```diff
@@ -0,0 +1,209 @@
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_FEMULTIVECTOR_DECL_HPP
+#define TPETRA_FEMULTIVECTOR_DECL_HPP
+
+/// \file Tpetra_FEMultiVector_decl.hpp
+/// \brief Declaration of the Tpetra::MultiVector class
+///
+/// If you want to use Tpetra::FEMultiVector, include "Tpetra_FEMultiVector.hpp"
+/// (a file which CMake generates and installs for you).  If you only want
+/// the declaration of Tpetra::FEMultiVector, include this file
+/// (Tpetra_FEMultiVector_decl.hpp).
+///
+
+#include "Tpetra_MultiVector_decl.hpp"
+
+namespace Tpetra {
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS
+  // forward declaration of MultiVector (declared later in this file)
+  template<class S, class LO, class GO, class N> class FEMultiVector;
+#endif // DOXYGEN_SHOULD_SKIP_THIS
+
+
+  template <class Scalar = ::Tpetra::Details::DefaultTypes::scalar_type,
+            class LocalOrdinal = ::Tpetra::Details::DefaultTypes::local_ordinal_type,
+            class GlobalOrdinal = ::Tpetra::Details::DefaultTypes::global_ordinal_type,
+            class Node = ::Tpetra::Details::DefaultTypes::node_type>
+  class FEMultiVector :
+    public MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>
+  {
+  private:
+    friend class MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
```

Comment:
Yes.  Because inheritance and templates don't work the way you think they do.  Or for that member, the way other people on C++ standards committees think they do.

Reply:
;-P

Reply:
What fails to build if it's not a friend?

Reply:
The compiler won't let you access any protected members in the base class.

Which is contrary to the "90's era C++" understanding of what "protected" means.

Reply:
@csiefer2 Did you try calling member functions with `this->`?  That's generally required with templated base classes.